org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/discovery/ZigBeeDiscoveryService.java

```diff
@@ -50,27 +57,64 @@
      */
     private final static int SEARCH_TIME = 60;
 
-    private final ZigBeeCoordinatorHandler coordinatorHandler;
-
     private DiscoveryServiceCallback discoveryServiceCallback;
 
-    public ZigBeeDiscoveryService(ZigBeeCoordinatorHandler coordinatorHandler) {
+    private final Set<ZigBeeCoordinatorHandler> coordinators = new CopyOnWriteArraySet<>();
+
+    private final Set<ZigBeeDiscoveryParticipant> participants = new CopyOnWriteArraySet<>();
+    private final Set<ThingTypeUID> supportedThingTypes = new CopyOnWriteArraySet<>();
+    private final Map<UID, ZigBeeNetworkNodeListener> registeredListeners = new ConcurrentHashMap<>();
+
+    public ZigBeeDiscoveryService() {
         super(SEARCH_TIME);
-        this.coordinatorHandler = coordinatorHandler;
-        logger.debug("Creating ZigBee discovery service for {}", coordinatorHandler.getThing().getUID());
+        logger.debug("Starting ZigBeeDiscoveryService");
     }
 
-    public void activate() {
-        logger.debug("Activating ZigBee discovery service for {}", coordinatorHandler.getThing().getUID());
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addZigBeeDiscoveryParticipant(ZigBeeDiscoveryParticipant participant) {
+        participants.add(participant);
+        supportedThingTypes.addAll(participant.getSupportedThingTypeUIDs());
+    }
 
-        coordinatorHandler.addNetworkNodeListener(this);
+    protected void removeZigBeeDiscoveryParticipant(ZigBeeDiscoveryParticipant participant) {
+        participants.remove(participant);
+        supportedThingTypes.clear();
+        for (ZigBeeDiscoveryParticipant currentParticipant : participants) {
+            supportedThingTypes.addAll(currentParticipant.getSupportedThingTypeUIDs());
+        }
     }
 
     @Override
     public void deactivate() {
-        logger.debug("Deactivating ZigBee discovery service for {}", coordinatorHandler.getThing().getUID());
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addZigBeeCoordinator(ZigBeeCoordinatorHandler coordinator) {
```

Comment:
Method name and parameter type do not match here - I'd suggest to rename ZigBeeCoordinatorHandler to ZigBeeCoordinator, see also my comment below.

Reply:
Are you ok if instead I rename the method to ``addZigBeeCoordinatorHandler```?  This class flows right through all aspects of the binding (as per the previous split of the dongles) and at this point I'd prefer not to change it. At some stage it might make sense to refactor out the name ```Coordinator``` as not all applications will be coordinators, but at again, that's not one for here ;)

Reply:
I am not asking you to rename or remove the class ZigBeeCoordinatorHandler - what I suggest is to introduce a new (additional) interface `ZigBeeCoordinator`, which is used here in the discovery service and which the handlers can implement (without the need of any further code changes in them).

Reply:
Sure - but I'd prefer to avoid doing this right now as this part of the code needs a major refactoring anyway. I'd prefer not to spend more time here which is completely wasted.